### DISCUSSION TITLE
How to horizontally center a table? align="CENTER" is not working for me

### DISCUSSION THREAD

**Discussion by DerekRobin** [py-pdf/fpdf2#1304] [Q&A] [answered]
I was wondering how to horizontally center a table on the page? I know about the `align` parameter that can be passed when instantiating a [table](https://github.com/py-pdf/fpdf2/blob/316b6e47a04c9f6a874dd8bc85e26b24d85e558d/fpdf/table.py#L58). However, for some reason passing `align="CENTER"` does not align the table to the center of the page.

I have included the code I am using to generate the table below from a pandas dataframe (df)
```python
df = df.applymap(str)
COLUMNS = [list(df)]  
ROWS = df.values.tolist()  
DATA = COLUMNS + ROWS
with self.table(
    borders_layout="NONE",
    headings_style=FontFace(emphasis="BOLD", color="#000000", fill_color="#ececec"),
    cell_fill_color="#ffffff",
    cell_fill_mode="EVEN_ROWS",
    line_height=self.font_size * 1.5,
    text_align=(
        "LEFT",
        "CENTER",
        "CENTER",
        "CENTER",
        "CENTER",
        "CENTER",
        "CENTER",
        "CENTER",
        "CENTER",
        "CENTER",
    ),
    col_widths=(190, 70, 70, 70, 70, 70, 70, 70),
    align="CENTER"
) as table:
    for i, data_row in enumerate(DATA):
        # some entries of data_row are conditionally modified here
        # in order to ensure the table has the correct data
        # the code within this function does not interface with any FPDF code
        data_row = modify_data_row()
       
        if i == 0:
            table.row(data_row)
        else:
            table.row(data_row, style=FontFace(color=color, fill_color=fill_color))
```

**Comment by Lucas-C** [comment, marked as answer]
Hi @DerekRobin

Thank you for the report.

Despite being mentioned in our documentation: https://py-pdf.github.io/fpdf2/Tables.html#setting-table-column-widths

> `align` can be passed to table() to set the table horizontal position relative to the page, when it's not using the full page width. It's centered by default.

...it seems that this never worked properly... 😅

I tested the following code with `fpdf2` version 2.7.0, where tables were originally introduced in this documentation was redacted:
```python
from fpdf import FPDF

TABLE_DATA = (
    ("First name", "Last name", "Age", "City"),
    ("Jules", "Smith", "34", "San Juan"),
    ("Mary", "Ramos", "45", "Orlando"),
    ("Carlson", "Banks", "19", "Los Angeles"),
    ("Lucas", "Cimon", "31", "Angers"),
)

pdf = FPDF()
pdf.add_page()
pdf.set_font("Times", size=16)
with pdf.table(col_widths=pdf.epw / 5, align="C") as table:
    for data_row in TABLE_DATA:
        row = table.row()
        for datum in data_row:
            row.cell(datum)
pdf.output("./test/table/table_with_fixed_col_width_and_align.pdf")
```

...and the table is **NOT** horizontally centered on the page!

Wooops.

I'm converting this discussion into a bug report, and I will look into fix it:
https://github.com/py-pdf/fpdf2/issues/1306

**Comment by DerekRobin** [comment]
Hey @Lucas-C, thanks for the fix!

I have tested the code and it seems to work for other tables in my document except for the tables created by the code I provided above, which is very strange.

I believe that it has something to do with the fact that the tables created by this code can have a max of 8 columns but may have less. Although this is just a hunch and isn't back by anything.

**Reply by Lucas-C** [reply]
Hi @DerekRobin

Oh OK, that's surprising.

I'd be curious to figure out the issue, but you will have to provide an autonomous [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example)

The code you providee is quite incomplete, and I cannot run as it is. What is the content of your DataDrame `df`? What is the value of `ROWS`? `self` is present so I can guess we are somehow in a method, maybe in a subclass of `FPDF`? What is `modify_data_row()`?

If you provide an **autonomous short Python script** that generates a PDF reproducing your problem, I'll be happy to have a look and further try to solve this out 🙂

**Reply by DerekRobin** [reply]
Hey @Lucas-C,

Here is a minimal reproducible example that shows how the behavior I am talking about. I have substituted in dummy data as I am unable to share the original data. Unfortunately I am unable to attach the example as a file so I have written it below and attached the CSV: [tables.csv](https://github.com/user-attachments/files/17851215/tables.csv)

```python
from fpdf import FPDF, FontFace
import pandas as pd

class MonthlyReport(FPDF): 
    
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        
    def page_break(self):
        self.write_html("<br>")

    
    def table_from_lists(self, df_list):
        self.set_font("Helvetica", "", 10)
        for df in df_list:
            # this can probably be moved elsewhere
            df = df.replace(" ", None)
            df = df.dropna(axis=1, how="all")
            df = df.applymap(str)
            df = df.replace("None", " ")
            COLUMNS = [list(df)]
            ROWS = df.values.tolist()
            DATA = COLUMNS + ROWS
            text_alignments = ["LEFT"]
            column_widths = [20]
            num_columns = len(df.columns) - 1
            with self.table(
                borders_layout="NONE",
                headings_style=FontFace(
                    emphasis="BOLD", color="#000000", fill_color="#ececec"
                ),
                cell_fill_color="#f3f3f3",
                cell_fill_mode="EVEN_ROWS",
                line_height=self.font_size * 1.5,
                text_align=("LEFT", "CENTER", "CENTER", "CENTER", "CENTER", "CENTER", "CENTER", "CENTER", "CENTER"),
                align="CENTER",
                col_widths=(10, 10, 10, 10, 10, 10, 10, 10, 10)
            ) as table:
                for i, data_row in enumerate(DATA):
                    if data_row[0] == "matched_compound":
                        data_row[0] = df.index.get_level_values(0)[0]
                    elif data_row[0] == "0":
                        data_row[0] = df.index.get_level_values(1)[i]

                    if data_row[0] == "expected":
                        color, fill_color = "#ffffff", "#174575"
                    elif data_row[0] == "additional":
                        color, fill_color = "#ffffff", "#7848b5"
                    elif data_row[0] == "unexpected":
                        color, fill_color = "#000000", "#d65aa5"
                    elif data_row[0] == "unknown":
                        color, fill_color = "#000000", "#ec9680"
                    else:
                        if i % 2 == 0:
                            color, fill_color = "#000000", "#f3f3f3"
                        else:
                            color, fill_color = "#000000", "#ffffff"

                    if i == 0:
                        table.row(data_row)
                    else:
                        table.row(
                            data_row, style=FontFace(color=color, fill_color=fill_color)
                        )
            self.page_break()

                        
if __name__ == "__main__":
    report = MonthlyReport()

    df=pd.read_csv("tables.csv")
    df["composition_group"] = pd.CategoricalIndex(
        df["composition_group"],
        ["expected", "additional", "unexpected", "unknown"],
        ordered=True,
    )
    df["Category"] = pd.CategoricalIndex(
        df["Category"], ["category1", "category2", "category3"], ordered=True
    )
    df_index = pd.MultiIndex.from_frame(
        df[["Category", "composition_group"]],
        names=["Category", "composition_group"],
    )
    df = df[
        [
            "matched_compound",
            "location1",
            "location2",
            "location3",
            "location4",
            "location5",
            "location6",
            "location7",
        ]
    ].set_index(df_index)
    df = df.sort_index(axis=0, level=0)
    
    df_list = [group for name, group in df.groupby(level=0)]
    
    report.add_page()
    report.table_from_lists(df_list)
    report.output(f"Example_Report.pdf")
```

**Reply by DerekRobin** [reply]
Upon providing a `width` to the table, it centers them but then the column widths are no longer respected. Maybe I was misunderstanding all along?

**Reply by Lucas-C** [reply]
Hi @DerekRobin

The solution is simply to pass `col_widths=10` instead of `col_widths=(10, 10, 10, 10, 10, 10, 10, 10, 10)` on line 37 in `table_from_lists()`, when calling `FPDF.table()`.

Why?
As I mentioned in #1306:
> Also, note that you will have to either specify a fixed column width (single number `col_widths` in the document unit) or a table `width=` in order for the table to not fill the page width.
> When `col_widths` is provided as an array, those values are not considered to be widths in the document `unit` but as "fractions" (meaning that `col_widths=(1, 1, 2)` is strictly equivalent to `col_widths=(25, 25, 50)`)

Now this is very tricky, and we should either:
* make it a lot more evident in our documentation ([docstring](https://py-pdf.github.io/fpdf2/fpdf/fpdf.html#fpdf.fpdf.FPDF.table) & [documentation page](https://py-pdf.github.io/fpdf2/Tables.html))
* change this behaviour, so that there are NOT 2 distinct interpretations of `col_widths`, depending on if it is a number or an iterable

What do you think about that @DerekRobin, @andersonhc, @gmischler? 🙂

**Reply by DerekRobin** [reply]
I think that the easiest solution is to make the documentation more evident. I believe beyond align not initially working, that this was where a lot my confusion came from.

Which is guess is to say that I think having two interpretations of an parameter is okay so long as that documented well

**Reply by Lucas-C** [reply]
Thank you for your answer @DerekRobin

I opened PR https://github.com/py-pdf/fpdf2/pull/1310 to improve the documentation

**Reply by DerekRobin** [reply]
Happy to help! And thanks for the amazing library, it's really helped automating report creation for my day job.

I have also closed this discussion, hope that's alright :)
